### PR TITLE
SWS vom 04.05.2021

### PR DESCRIPTION
**Pull request by ZeroPointMax**

### CODE REVIEW

**Comment by Steve2955** [comment]
## Command results
<details>
<summary>Details: </summary>

<details>
<summary><em>Run doctoc</em></summary>

```Shell
changed:
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/ABWL-LEDER/README.md
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/SWS-KUSCHE/README.md
unchanged:
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/ASM-GUENTHER/README.md
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/DBS-DORENDORF/README.md
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/PRO-KASCHE/README.md
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/SWE-KASCHE/README.md
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/EMB-GUENTHER/README.md
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/OOP-CASSELT/README.md
```



</details>

</details>

## Changed files
<details>
<summary>Changed 2 files: </summary>

- ABWL-LEDER/README.md
- SWS-KUSCHE/README.md

</details>

<hr>

[:octocat: Repo](https://github.com/technote-space/toc-generator) | [:memo: Issues](https://github.com/technote-space/toc-generator/issues) | [:department_store: Marketplace](https://github.com/marketplace/actions/toc-generator)

**Comment by Steve2955** [comment]
## Command results
<details>
<summary>Details: </summary>

<details>
<summary><em>Run doctoc</em></summary>

```Shell
changed:
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/SWS-KUSCHE/README.md
unchanged:
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/ABWL-LEDER/README.md
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/ASM-GUENTHER/README.md
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/DBS-DORENDORF/README.md
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/PRO-KASCHE/README.md
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/SWE-KASCHE/README.md
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/EMB-GUENTHER/README.md
  - /home/runner/work/dhge-pi19-sem4/dhge-pi19-sem4/OOP-CASSELT/README.md
```



</details>

</details>

## Changed files
<details>
<summary>Changed file: </summary>

- SWS-KUSCHE/README.md

</details>

<hr>

[:octocat: Repo](https://github.com/technote-space/toc-generator) | [:memo: Issues](https://github.com/technote-space/toc-generator/issues) | [:department_store: Marketplace](https://github.com/marketplace/actions/toc-generator)

File: SWS-KUSCHE/README.md [hunk 1]
```diff
@@ -942,3 +944,133 @@ Es gibt verschiedene Arten der Disk Encryption:
...
+  - ungeprüfte Input-Längen, v.a. bei Strings
+  - Kombination von fixen Arrays und zu großen dynamischen Daten
+  - Array-Grenzen und Pointer-Arithmetik; Off-by-one-Fehler
+  - fehlende String-Terminatoren
+  - Integer Overflow und malloc-Größenberechnungen
+  - ``free``-Fehler
+  - Return auf Pointer auf lokale Daten; nicht initialisierte Daten
+- bekannt unsichere Funktionen: ``gets``, ``scanf``, ``strcpy``, ``strncpy``, ``strcat``
+- finden von Bugs:
+  - Lesen von Quellcode oder Reverse Engineering an Absturzstellen und Stellen, die durch Security Fixes geändert wurden+
```
Comment (Steve2955) [inline SWS-KUSCHE/README.md:982]:
```suggestion
  - Lesen von Quellcode oder Reverse Engineering an Absturzstellen und Stellen, die durch Security Fixes geändert wurden
```

File: SWS-KUSCHE/README.md [hunk 2]
```diff
@@ -942,3 +944,133 @@ Es gibt verschiedene Arten der Disk Encryption:
...
+  - enthalten Codefragment, das ausgeführt werden soll
+
+**Return-Oriented Programming**
+
+- Manipulation des Stacks, sodass Returnadresse auf ein bestehendes Codestück zeigt, v.a. in Libraries, z.B. ``system``
+- kein eingeschmuggelter Payload, sondern Nutzung des bereits vorhandenem Code
+
+**gefährliches Lesen**
+
+- leaken von Daten
```
Comment (Steve2955) [inline SWS-KUSCHE/README.md:996]:
```suggestion
- Leaken von Daten
```

File: SWS-KUSCHE/README.md [hunk 3]
```diff
@@ -942,3 +944,133 @@ Es gibt verschiedene Arten der Disk Encryption:
...
+- Sandboxing
+  - Filterung aller System-Calls zur Einschränkung von Zugriffen
+  - Urformen: ``chroot``, ``seccomp``
+  - erweiterte Formen: Role Based Access Control (RBAC); Capabilities (SELinux, AppArmor); ``seccomp-bpf``
+  - hilft nicht direkt gegen Überschreiber, aber mildert die Folgen
+- Brute-Force-Bremse: automatischen Neuversuch verzögern
+- Trusted Path Execution:
+  - Liste von Programm-Directories $\rightarrow$ *nur* diese Dirs sind ausführbar, aber *nie* schreibbar
+  - unterbindet Ausführung heruntergeladener Dateien
+- seit kurzem: ``vtable``-Verfication: Prüfung der Vtables zur Laufzeit mittels Prüfsumme; GCC-Option
```
Comment (Steve2955) [inline SWS-KUSCHE/README.md:1034]:
```suggestion
- seit Kurzem: ``vtable``-Verification: Prüfung der Vtables zur Laufzeit mittels Prüfsumme; GCC-Option
```

File: SWS-KUSCHE/README.md [hunk 4]
```diff
@@ -942,3 +944,133 @@ Es gibt verschiedene Arten der Disk Encryption:
...
+  - speichert bei jedem Call die Return-Adresse, vergleicht diese
+  - Interrupt bei Korruption
+  - schützt nur Return-Adressen, aber auch nicht lückenlos
+- Execute Protection (NX-Bit, DEP, W^X):<!--letzteres: Write XOR Execute-->
+  - Speicher darf nie gleichzeitig schreibbar **und** ausführbar sein
+  - CPU-Feature (d.h. sehr schnell)
+  - nicht vorhanden auf Embedded Systemen
+  - hilft nicht gegen Return Oriented Programming
+  - Skripte mit **JIT-Kompiler** müssen teilweise **ohne Execute Protection** ausgeführt werden
+- Adress Space Layout Randomization (ASLR):
```
Comment (Steve2955) [inline SWS-KUSCHE/README.md:1021]:
```suggestion
- Address Space Layout Randomization (ASLR):
```

File: SWS-KUSCHE/README.md [hunk 5]
```diff
@@ -942,3 +944,133 @@ Es gibt verschiedene Arten der Disk Encryption:
...
+  - unterbindet Ausführung heruntergeladener Dateien
+- seit kurzem: ``vtable``-Verfication: Prüfung der Vtables zur Laufzeit mittels Prüfsumme; GCC-Option
+- Programmier-Richtlinien; Textsuche nach gefährlichen Konstrukten; viele Compilerwarnungen; statische Programmanalyse ("Linting"); Code-Reviews
+  - statische Programmanalyse untersucht für jede Variable den möglichen Wertebereich (auch ``NULL``)
+    - merken sich z.B. für Pointer, ob ``free`` / ``delete`` aufgerufen wurde oder mehrere Pointer auf dieselbe Adresse zeigen
+  - Speicherzugriffschecker: Compiler-basiert, z.B. ASAN im gcc und llvm; Code Instrumenter, z.B. Purify; interpretierende Tools, z.B. Valgrind
+    - prüfen Array- und Pointer-Zugriffe; führen Buch über allokierte Blöcke und uninitialisierte Speicherbereiche
+    - erfordert Verständnis, um zu erkennen, welches Tool welche Fehler erkennen kann
+  - Testen
+    - kreative, menschliche Tester; spezielle Pen-Tester; Fuzzing-Programme
```
Comment (Steve2955) [inline SWS-KUSCHE/README.md:1042]:
```suggestion
  - Testen: kreative, menschliche Tester; spezielle Pen-Tester; Fuzzing-Programme
```

**Review by Steve2955 — APPROVED** [review]

File: SWS-KUSCHE/README.md (hunk 4 above)
Comment (ZeroPointMax) [reply SWS-KUSCHE/README.md:1021]: Die unterschiedliche Rechtschreibung zwischen Deutsch und Englisch macht mich fertig :D

**Review by nukerxy — APPROVED** [review]

**Pull request merged**